### PR TITLE
E2e protocol ricardo 20231025 login sync

### PR DESCRIPTION
**Pull request by corrideat**

### CODE REVIEW

**Comment by Silver-IT** [comment]
Having checked this PR mainly focusing on the user flow, I could see several issues.

**Issue 1**
- Steps to reproduce
  - [Tab1] user1 signs up, and creates a group
  - [Tab2] user2 joins the group, and closes the tab
  - [Tab1] user1 kicks the u2 from the group
  - [Tab2] Opens 'http://localhost:8000'. Got 5 errors in the console
  - [Tab1] Got error in console too.
- Screenshot
  - Error in Tab1
image
  - 5 Errors in Tab2
image
image
image
image
- What should happen
  - No errors should be happening. And the important thing is that there MUST be no errors in Tab1.
  - `gi.actions/group/join` shouldn't be called. But it is called twice.
- DeadLock Output
  - [Tab1] {}
  - [Tab2] {}

**Comment by Silver-IT** [comment]
**Issue 2**
- Steps to reproduce
  - [Tab1] user1 signs up, and creates a group
  - [Tab2] user2 joins the group, and closes the tab
  - [Tab1] user1 kicks the u2 from the group
  - [Tab3] user2 logs in, and got 5 errors.
- Screenshot
  - 5 Errors in Tab2
image
image
image
image
image
- What should happen
  - No errors should be happening.
  - `gi.actions/group/join` should be called ONLY once. But it is called twice.
- DeadLock Output
  - [Tab1] {}
  - [Tab2] {}

**Comment by Silver-IT** [comment]
**Issue 3**
- Steps to reproduce
  - [Tab1] user1 signs up, and creates a group.
  - [Tab2] user2 joins the group.
  - [Tab1] user1 kicks the u2 from the group. Got 1 error in console.
  - [Tab2] user2 joins the group again with the initial invite link.
  - [Tab2] Wrong work flow to join. 6 Errors in console. Seems there are many redundant function calls.
  - [Tab1] Many logs in console than expected. Seems there are too many function calls. And duplicated notifications.
- Screenshot
  - Error in Tab1 (Step 3)
image
  - Errors in Tab2 (Step 5)
image
image
  - Duplicated notifications in Tab1 (Step 6)
image
- What should happen
  - No errors should be happening.
  - No duplicated notifications in Tab1
  - `gi.actions/group/join` should be called ONLY once. But it is called 4 times.
- DeadLock Output
  - [Tab1]
image
  - [Tab2] {}

**Comment by Silver-IT** [comment]
**Issue 4**
- Steps to reproduce
  - [Tab1] user1 signs up, and creates a group.
  - [Tab2] user2 joins the group.
  - [Tab3] user3 joins the group, and closes the tab.
  - [Tab2] user2 creates a proposal to remove user3 from the group.
  - [Tab1] user1 approves the proposal, and kicks user3 from the group. Got an error in the notifications.
  - [Tab4] user3 tries to log in, and 8 errors BANG.
- Screenshot
  - Error in Tab1 (Step 5)
(user1 = alexjin, user2 = greg, user3 = andrea)
image
  - Errors in Tab4 (Step 6)
image
image
image
image
image
image
image
image
- What should happen
  - No errors should be happening.
  - `gi.actions/group/join` should be called ONLY once. But it is called twice in Tab4 (Step 6).
- DeadLock Output
  - [Tab1]
image
  - [Tab2] {}
  - [Tab4] {}

**Comment by Silver-IT** [comment]
**Issue 5**
- Steps to reproduce
  - [Tab1] user1 signs up, and creates a group.
  - [Tab2] user2 joins the group.
  - [Tab3] user1 logs in.
  - [Tab4] user2 logs in.
  - [Tab1] user1 kicks user2 from the group.
  - [Tab1], [Tab2], [Tab4] Errors BANG.
- Screenshot
  - Error2 in Tab1 (Step 6)
(user1 = alexjin, user2 = greg)
image
  - Errors in Tab2 (Step 6)
image
  - Errors in Tab4 (Step 6)  
image
- What should happen
  - No errors should be happening.
- DeadLock Output
  - [Tab1] {}
  - [Tab2] {}
  - [Tab3] {}
  - [Tab4] {}

**Comment by corrideat** [comment]
Thanks again, @Silver-IT!

As discussed:

1. I've made side-effects mostly work as they did earlier
2. I've added a check on `syncContractAndWatchKeys` to avoid syncing contracts that have been subscribed to

In addition, I've implemented fixes for all of the issues reported here, except issue 3. I'll continue working on this until issue 3 is resolved.

Regarding issue 1, there is the double sync, that I've tracked down to the login functionality. I'll work on addressing this as well (but it seems to be the case that implementing the `join|leaveGroup` actions would help solving it too).

@taoeffect: You'll probably be pleased to see that I've mostly removed `sync` from `recreateEvent`, something you weren't entirely happy with. There are some vestiges of it that remain, because _sometimes_ the `pubsub` functions don't work as expected and some events are never received. Once this is addressed in `pubsub`, the `syncContract` added to `publishEvent` can be removed.

**Comment by corrideat** [comment]
@Silver-IT I've made another commit addressing issue 3. The solution so far is using `sbp('okTurtles.data/set', 'JOINING_CHATROOM-' + params.contractID, ?)` to avoid removing the chatroom contract when joining.

Besides that, two issues remain.

One is at _leaving_; specifically, when someone else leaves a chatroom we have left:

To write to a contract, we need to be subscribed to it. This is done in `encryptedAction`, but for contracts that call `/remove` in a side-effect, this is essentially pointless. This is an error.

The second issue is at _joining_. You'll see that the `/join` action calls `/sync`. This is 'unnecessary', as `encryptedAction` already does this. However, similarly to what happened above, the first `/sync` will be pointless because of the side-effect removing the contract, requiring a second call to sync that sets the `JOINING_CHATROOM-${cID}` key.

I think that we need to either adapt `encryptedAction` to take pre- and post-hooks (we could call them, e.g., prologue and epilogue) to handle this, or somehow re-think how `/remove` gets called (I think that one thing that would work is doing reference counting, but there may be other options too). These are not mutually exclusive, so we could very well do both.

Related to this, there is a third issue, which is a variant of issue 3:

* rejoin fails 
  - [tab1] u1: create group
  - [tab2] u2: join using invite link
  - [tab2] u2: close tab / navigate away
  - [tab1] u1: remove u2
  - [tab2] u2: open tab with invite link

To recap, apart from the issues mentioned, which are related to removing contracts, there is the double-sync issue I mentioned earlier.

Although double-syncs are a concern, I'll focus now on trying to address the actual issues that happen from missing contract states, which are errors.

Do you have a suggestion on how we could call `/remove` in an elegant way?

File: frontend/model/contracts/group.js
Comment (Silver-IT) [inline frontend/model/contracts/group.js]: We don't need to call `gi.contracts/group/leaveGroup` when the member has joined the group after this action.
So it would be necessary to check the group contract profile state in this invocation.
Please reference [my PR](https://github.com/okTurtles/group-income/pull/1728) for this.

```
sideEffect ({ data, meta, contractID }, { state, getters }) {
  const rootState = sbp('state/vuex/state')
  const rootGetters = sbp('state/vuex/getters')
  const contracts = rootState.contracts || {}
  const { username } = rootState.loggedIn

  if (data.member === username) {
    // If this member is re-joining the group, ignore the rest
    // so the member doesn't remove themself again.
    sbp('chelonia/queueInvocation', contractID, async () => {
      if (rootState[contractID].profiles?.[username]?.status === PROFILE_STATUS.REMOVED) {
        // NOTE: should remove archived data from IndexedStorage
        //       regarding the current group (proposals, payments)
        await sbp('gi.contracts/group/removeArchivedProposals', contractID)
        await sbp('gi.contracts/group/removeArchivedPayments', contractID)
```

File: frontend/model/contracts/group.js
Comment (Silver-IT) [inline frontend/model/contracts/group.js:324]: Inside `group` contract, it doesn't make sense to decide the next operation depending on the `chatroom` contract state.

**Review by Silver-IT — CHANGES_REQUESTED** [review]
Nice work, @corrideat. I did a very partial review, and left a just two comments.
I can see many updates in contract parts (model/contracts, controller/actions) but I don't think most of them are needed. I have almost finished my PR #1728 and I want you to have even a brief look at it. I fixed most of the login issues with the smaller changes than this.

Also will check the 5 Issues (described above) again, and left comments here.

**Comment by Silver-IT** [comment]
**Issue 1** (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784531132) This issue is not fixed yet.
image
This screenshot shows the console logs filtered by `gi.actions/group` on Tab2 in Step 5 .

**Problem: three errors and 2 redundant `gi.actions/group/join` function calls**

**Comment by Silver-IT** [comment]
**Issue 2** (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784557201) This issue is not fixed yet.
image
This screenshot shows the console logs filtered by `gi.actions/group/join` on Tab3 in Step 4.

**Problem: 1 redundant `gi.actions/group/join` function call.**

**Comment by Silver-IT** [comment]
Issue 3 (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784603749) seems not to be fixed yet.
image
image
image

**Problem: two errors in console**

File: shared/domains/chelonia/GIMessage.js
Comment (taoeffect) [inline shared/domains/chelonia/GIMessage.js]: No `console.warn` is needed here?

File: shared/domains/chelonia/GIMessage.js
Comment (taoeffect) [inline shared/domains/chelonia/GIMessage.js:411]: Why was this changed?

File: frontend/views/pages/PendingApproval.vue
Comment (taoeffect) [inline frontend/views/pages/PendingApproval.vue:40]: This change changes the meaning of this computed property.

The computed property is named `ourGroupProfile`, which implies it returns our group profile. But now this no longer happens, instead it returns a boolean.

So either the condition needs to be removed or the name of the computed property needs to be changed.

**Review by taoeffect — CHANGES_REQUESTED** [review]
Preliminary review

File: frontend/common/errors.js
Comment (taoeffect) [inline frontend/common/errors.js]: I think this is a good time to copy over that utility function from Chelonia to define these using 1-liners...

File: shared/domains/chelonia/GIMessage.js
Comment (corrideat) [reply shared/domains/chelonia/GIMessage.js:411]: Because I removed a bunch of the repetitive code in `chelonia.js` that was looking up the previous HEAD, which is now done by `publishEvent`. Since the previous HEAD could now momentarily be null, the check as it was would fail. Note that either condition is correct (if the previous HEAD were set).

File: shared/domains/chelonia/GIMessage.js
Comment (corrideat) [reply shared/domains/chelonia/GIMessage.js]: This is doing the same as was it was doing earlier, more or less. There could be a `console.warn`, but that is already done elsewhere and it feels repetitive to warn multiple times.

File: frontend/views/pages/PendingApproval.vue
Comment (corrideat) [reply frontend/views/pages/PendingApproval.vue:40]: Agreed.

File: frontend/common/errors.js
Comment (corrideat) [reply frontend/common/errors.js]: Agreed.

File: shared/domains/chelonia/internals.js
Comment (taoeffect) [inline shared/domains/chelonia/internals.js]: So I agree that we should address this in this PR... one way to go about it is to make sure the pubsub is connected before `'chelonia/contract/sync'` does the sync.

We might also want to verify that we are in fact connected to the pubsub server before allowing messages to be sent.

File: frontend/model/contracts/group.js
Comment (taoeffect) [inline frontend/model/contracts/group.js]: This line seems to be removing many profile fields... that seems wrong.

File: frontend/model/contracts/group.js
Comment (taoeffect) [inline frontend/model/contracts/group.js:303]: These are good comments 👍

File: frontend/model/contracts/group.js
Comment (taoeffect) [inline frontend/model/contracts/group.js:312]: Does `innerSigningContractID = null` need to be specified? If yes, could you add a comment explaining why?

File: frontend/model/contracts/group.js
Comment (corrideat) [reply frontend/model/contracts/group.js]: There isn't much in the profile _and_ this is for removed members. Would you prefer `{ ...profile, status: PROFILE_STATUS.REMOVED }` instead?

File: frontend/model/contracts/group.js
Comment (corrideat) [reply frontend/model/contracts/group.js:312]: Yes, it needs to be specified or else it'll be set to the identity CSK.

File: frontend/model/contracts/group.js
Comment (taoeffect) [reply frontend/model/contracts/group.js]: Yes, I think that would be much more robust.

File: frontend/model/contracts/group.js
Comment (taoeffect) [inline frontend/model/contracts/group.js:327]: ~~This `.cause` thing doesn't seem to be a thing? `GIErrorMissingSigningKeyError` sets `.name` to `GIErrorMissingSigningKeyError`.~~

EDIT: nevermind, it's set in `actions/utils.js`

File: shared/domains/chelonia/utils.js
Comment (taoeffect) [inline shared/domains/chelonia/utils.js:490]: What's going on here with `OP_ATOMIC`? It seems like if it's an atomic op with 1 operation it's for some reason modifying the message that's being sent to remove the outer `OP_ATOMIC` wrapper... but that's wrong. Why is it modifying what the user wants?

Comment (taoeffect) [inline shared/domains/chelonia/utils.js]: What's going on here? Are you sure this is right? Is `? (newRawOpV: any)[0]` correct?

Why is this being done? Is it rewriting what the user is sending? We should probably keep the developer intent and not modify messages the developer is sending.

File: shared/domains/chelonia/internals.js
Comment (taoeffect) [inline shared/domains/chelonia/internals.js]: Why is this wrapping the result in `Promise.resolve`? What if an error occurred? What's the point of later checking `.then` and `Promise.allSettled` if they're guaranteed to settled because of the `Promise.resolve`?

Or is that not how it works? If so please add a comment explaining because this is confusing...

Comment (corrideat) [reply shared/domains/chelonia/internals.js]: Agreed

File: shared/domains/chelonia/GIMessage.js
Comment (taoeffect) [reply shared/domains/chelonia/GIMessage.js]: Then a comment about that should be here...

File: shared/domains/chelonia/utils.js
Comment (corrideat) [reply shared/domains/chelonia/utils.js]: Reverted

**Comment by Silver-IT** [comment]
- Still existing issues:
  - **Issue 1**(https://github.com/okTurtles/group-income/pull/1780#issuecomment-1797865986)
  - **Issue 2** (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1797876669)
  - **Issue 3** (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1797889651)
  - **Issue 4** (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784699012)
  Still two function `gi.actions/group/join` calls

**Comment by Silver-IT** [comment]
@corrideat, @taoeffect. I have just checked the test scenario.
### Still existing issues in terms of workflow
  - [**Issue 1**](https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784531132)
  Problem:
    - user2 has successfully logged in.
    - One error occurred.
image

- [**Issue 2**](https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784557201)
Problem:
  - Two errors occurred.
  - `gi.actions/group/join` is called twice. It should be called only once.
image
image
image

- [**Issue 3**](https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784603749)
Problem:
  - Error in [Tab2] after being kicked from the group. (Not before to try join)
  - Three `gi.actions/group/join` calls. Only two calls are needed.
image
image

- [**Issue 4**](https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784699012)
Problem: 
  - 4 errors in console in Tab4
  - `gi.actions/group/join` is called twice. It should be called only once.
image
image
- [**Issue 5**](https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784711834)
Seems fine.

(user1 = alexjin, user2 = greg, user3 = andrea)


IMHO, I don't approve the many updates in `model/contracts` and `controllers/actions`. Especially, I can't approve the way to leave chatroom when leaving group.

File: shared/domains/chelonia/internals.js
Comment (taoeffect) [inline shared/domains/chelonia/internals.js:972]: Is this logging still needed?

**Review by taoeffect — APPROVED** [review]
This is looking really solid now, thanks @Silver-IT for finding those issues and @corrideat for figuring out the fixes! I think it's ready to merge into `e2e-protocol` but with also double-check with @Silver-IT to see if he can confirm that the issues he found have been fixed.

**Pull request merged**
